APIv2(DeliverMax): add alias for Amount in Payment transactions

Bronek opened the pull request:
<!--
This PR template helps you to write a good pull request description.
Please feel free to include additional useful information even beyond what is requested below.
-->

## Add `DeliverMax` alias to `Amount`, remove `Amount` from API ver 2.

<!--
Please include a summary of the changes.
This may be a direct input to the release notes.
If too broad, please consider splitting into multiple PRs.
If a relevant task or issue, please link it here.
-->

### Context of Change

Resolve https://github.com/XRPLF/rippled/issues/3484 

### Type of Change

<!--
Please check [x] relevant options, delete irrelevant ones.
-->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [ ] Tests (You added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation Updates
- [ ] Release

Bronek commented:
I realised the old version of this draft was updating `meta` json element (which is where `delivered_amount` is), but it should instead update `transaction` json element (not necessarily under this name - we are unfortunately inconsistent), because that's where `Amount` is. New version of this PR fixes this.

Bronek commented:
Also found that I missed proposed transactions and proposed account transaction, this is fixed now. I am considering some refactoring of my changes in `NetworkOPs.cpp` to improve cohesion and reduce coupling - i.e. the way I added virtually identical non-trivial block of code in several functions, while missing proposed transactions. So that will be probably next commit and then of course badly needed unit tests.

Bronek commented:
Note, as seen in `Add test for TransactionSign.cpp` commit, using DeliverMax instead of `Amount` does not affect the transaction signature: both tests `"Invalid Amount in submit_multisigned Payment."` and `"Invalid DeliverMax in submit_multisigned Payment."` have the same signature. Also the error in the latter test refers to `Amount` and not `DeliverMax` field. This is consequence of the fact that `DeliverMax` is an alias to `Amount`.

arihantkothari commented on src/ripple/app/misc/NetworkOPs.cpp:
```diff
@@ -599,10 +601,12 @@ class NetworkOPsImp final : public NetworkOPs
         bool validated,
-        std::shared_ptr<ReadView const> const& ledger);
+        std::optional<std::reference_wrapper<TxMeta const>> meta,
+        int apiVersion);
```
Since `apiVersion` will always be a positive integer and` Context::apiVersion` is an `unsigned int` [[source]](https://github.com/XRPLF/rippled/blob/3e08c390f5f18b308ace0dbdbb27fa02e49141fa/src/ripple/rpc/Context.h#L50)
Isn't it better to make `apiVersion` in this PR an `unsigned int` as well or am I missing something ?

intelliot commented:
@Bronek - can you confirm this PR is now ready for review?

Bronek commented:
> @Bronek - can you confirm this PR is now ready for review?

Yes, it is.

mDuo13 commented on API-CHANGELOG.md:
```diff
@@ -149,6 +149,15 @@ Since `api_version` 2 is in "beta", breaking changes to V2 can currently be made
 
+##### In progress
+
+- In `Payment` transaction type, JSON RPC field `Amount` is renamed to `DeliverMax`. To enable smooth client transition, `Amount` is still handled, as described below:
+  - On JSON RPC input (e.g. `submit_multisigned` etc. methods), `Amount` is recognized as an alias to `DeliverMax` for both API version 1 and version 2 clients.
+  - On JSON RPC input, submitting both `Amount` and `DeliverMax` fields is allowed _only_ if they are identical; otherwise such input is rejected with `rpcINVALID_PARAMS` error.
+  - On JSON RPC output (e.g. `subscribe`, `account_tx` etc. methods), `DeliverMax` is present in both API version 1 and version 2.
+  - On JSON RPC output, `Amount` is only present in API version 1 and _not_ in version 2.
```
JSON-RPC should have a hyphen, although I think the changes _should_ apply to both WebSocket and JSON-RPC (and likely do, based on what I see in the code here). Frequently people use the term "RPC" to encompass both the JSON-RPC and WebSocket APIs, although I tend to avoid such usage because it can cause such confusion.

intelliot replied:
Are we doing JSON-RPC over WebSocket? For HTTP JSON-RPC, maybe we should say "HTTP" 🤔

intelliot commented:
@arihantkothari to look over changes made after prior review

intelliot commented:
note: @ximinez is reviewing and will share an update shortly

ximinez reviewed (changes requested):
I've gone ahead an made most or all of the suggested changes on a branch, mostly to make sure the changed `MultivarJson` initialization worked like I expected. https://github.com/Bronek/rippled/compare/feature/partial-payment-fieldname...ximinez:rippled:bk-4733-fieldname-2

Feel free to `git cherry-pick` the commit(s), or copy/paste the changes you want, or do your own thing.

ckeshava commented on src/test/rpc/AccountTx_test.cpp:
```diff
@@ -119,14 +119,22 @@ class AccountTx_test : public beast::unit_test::suite
-                  [jss::TransactionType] == jss::Payment);
+                  [jss::TransactionType] == jss::Payment) &&
+                (j[jss::result][jss::transactions][1u][jss::tx]
+                  [jss::DeliverMax] == "10000000010") &&
```
I'm aware that this value has been used in `AccountTx` and `Subscribe` unit tests, but I'm not able to piece it together.

Bronek replied:
it's a 10'000 XRP transaction which converts to 10'000'000'000 drops, plus minimum fee of 10 drops

ckeshava commented on src/test/rpc/AccountTx_test.cpp:
```diff
@@ -119,14 +119,22 @@ class AccountTx_test : public beast::unit_test::suite
+                  [jss::TransactionType] == jss::Payment) &&
+                (j[jss::result][jss::transactions][1u][jss::tx]
+                  [jss::DeliverMax] == "10000000010") &&
+                ((apiVersion > 1 &&
+                  !j[jss::result][jss::transactions][1u][jss::tx].isMember(
+                      jss::Amount)) ||
+                 (j[jss::result][jss::transactions][1u][jss::tx][jss::Amount] ==
+                  j[jss::result][jss::transactions][1u][jss::tx]
+                   [jss::DeliverMax]));
```
good catch!

Bronek replied:
Thanks, will fix !

Bronek commented on src/ripple/app/ledger/impl/LedgerToJson.cpp:
```diff
@@ -23,6 +23,7 @@
 #include <ripple/basics/base_uint.h>
 #include <ripple/core/Pg.h>
 #include <ripple/rpc/Context.h>
+#include <ripple/rpc/DeliverMax.h>
```
Thanks, done; I was confused by some aspects of the design and so followed the "better to be consistent" rule here.

ximinez replied:
> Thanks, done; I was confused by some aspects of the design and so followed the "better to be consistent" rule here.

_Usually_ that's the way to go, but improving levelization is one of my ongoing goals, so when I catch it, I like to do something about it.

Bronek commented on src/ripple/app/misc/NetworkOPs.cpp:
```diff
@@ -599,10 +601,12 @@ class NetworkOPsImp final : public NetworkOPs
 
     Json::Value
     transJson(
-        const STTx& transaction,
+        std::shared_ptr<STTx const> const& transaction,
```
That's because this function now calls `insertDeliveredAmount` which takes `std::shared_ptr<STTx const>` ; moving this call here helped to clean the design of this part - basically not other part of this file now touches the transaction JSON output, which IMO is improvement. I did consider changing the signature of `insertDeliveredAmount` to take `const STTx&` instead of `shared_ptr` so I do not need `shared_ptr` here, but realised that would cause large unnecessary churn in other unrelated parts, and that's not a right PR for this kind of change.

ximinez replied:
> That's because this function now calls `insertDeliveredAmount` 

Ah! That makes sense!

Bronek commented on src/ripple/app/misc/NetworkOPs.cpp:
```diff
@@ -2744,7 +2748,9 @@ NetworkOPsImp::pubProposedTransaction(
     TER result)
 {
-    Json::Value jvObj = transJson(*transaction, result, false, ledger);
+    MultiApiJson jvObj(
+        {transJson(transaction, result, ledger, false, std::nullopt, 1),
+         transJson(transaction, result, ledger, false, std::nullopt, 2)});
```
We are building different JSONs here, for different API versions. One still has `Amount` field and the other does not. In another PR (to unify transaction JSON output - which is built on top of this one) the differences are much larger - different name of JSON element for the transaction data, different location of `hash` element etc. In principle it is possible to generate one JSON element and then transform it from one version to another, but that would be very brittle. I think the best we can do is to do some kind of memoization, so building the transaction JSON second time takes fewer cycles.

Bronek replied:
... needless to say, the way JSON population currently works (any part of code can do JSON), such memoization would cause not a small amount of churn, which I'd rather avoid.

Bronek replied:
oh I just looked at your branch. Good idea with changing the type of `transJson` result, I think I am going to copy it.

Bronek commented on src/ripple/app/misc/NetworkOPs.cpp:
```diff
@@ -3083,10 +3090,12 @@ NetworkOPsImp::getLocalTxCount()
-    const STTx& transaction,
+    std::shared_ptr<STTx const> const& transaction,
     TER result,
+    std::shared_ptr<ReadView const> const& ledger,
     bool validated,
```
TBH do not remember. I suppose I was hoping to make something easier to read or less brittle, but looking at it now I do not see any benefit; switched them back.

Bronek commented on src/test/rpc/JSONRPC_test.cpp:
```diff
@@ -1905,6 +1953,72 @@ static constexpr TxnTestData txnTestArray[] = {
        "A Signer may not be the transaction's Account "
        "(rHb9CJAWyB4rj91VRWn96DkukG4bwdtyTh)."}}},
 
+    {"Empty Signers array in submit_multisigned, use DeliverMax",
```
looking for specific error report which does **not** report missing `Amount`

Bronek commented on src/test/rpc/JSONRPC_test.cpp:
```diff
@@ -1035,6 +1035,26 @@ static constexpr TxnTestData txnTestArray[] = {
        "Missing field 'tx_json.Destination'.",
        "Missing field 'tx_json.Destination'."}}},
 
+    {"Missing 'Destination' in sign_for, use DeliverMax",
```
looking for specific error which does **not** report missing `Amount`

ckeshava replied:
I'm sorry, I don't understand this, what do you mean?

Bronek commented on src/ripple/basics/MultivarJson.h:
```diff
@@ -0,0 +1,70 @@
+#ifndef RIPPLE_BASICS_MULTIVARJSON_H_INCLUDED
+#define RIPPLE_BASICS_MULTIVARJSON_H_INCLUDED
+
+#include <ripple/json/json_value.h>
```
will do, thanks

Bronek commented on src/ripple/app/misc/NetworkOPs.cpp:
```diff
@@ -3294,30 +3312,37 @@ NetworkOPsImp::pubAccountTransaction(
+            {transJson(stTxn, trResult, ledger, true, metaRef, 1),
+             transJson(stTxn, trResult, ledger, true, metaRef, 2)});
 
+        for (InfoSub::ref isrListener : notify)
         {
-            auto const& meta = transaction.getMeta();
-
-            jvObj[jss::meta] = meta.getJson(JsonOptions::none);
-            RPC::insertDeliveredAmount(jvObj[jss::meta], *ledger, stTxn, meta);
+            isrListener->send(
+                jvObj.select(apiVersionSelector(isrListener->getApiVersion())),
+                true);
         }
 
-        for (InfoSub::ref isrListener : notify)
-            isrListener->send(jvObj, true);
-
         if (last)
-            jvObj[jss::account_history_boundary] = true;
+            jvObj.set(jss::account_history_boundary, true);
 
-        assert(!jvObj.isMember(jss::account_history_tx_stream));
+        assert(!jvObj.val[0].isMember(jss::account_history_tx_stream));
+        assert(!jvObj.val[1].isMember(jss::account_history_tx_stream));
```
Good idea, will do it. Note that however that it's no longer a boolean kind of a query. We want to enforce that "none of the objects have this member", as opposed to e.g. "some of the objects do not have this member".

Bronek commented on src/ripple/app/misc/NetworkOPs.cpp:
```diff
@@ -3294,30 +3312,37 @@ NetworkOPsImp::pubAccountTransaction(
-            info.sink_->send(jvObj, true);
+                jvObj.set(jss::account_history_tx_first, true);
+
+            jvObj.set(jss::account_history_tx_index, index->forwardTxIndex_);
+            index->forwardTxIndex_ += 1;
```
Thanks, will change back

Bronek commented on src/ripple/net/impl/InfoSub.cpp:
```diff
@@ -136,4 +136,16 @@ InfoSub::getRequest()
     return request_;
 }
 
+void
+InfoSub::setApiVersion(unsigned int apiVersion)
+{
+    this->apiVersion = apiVersion;
+}
+
+unsigned int
+InfoSub::getApiVersion() const noexcept
+{
+    return this->apiVersion;
+}
```
FWIW I have gotten into habit of writing `this->` from writing of template classes, where names resolve differently if you forget `this->` . Not needed here obviously, will remove it.

Bronek commented on src/ripple/rpc/impl/TransactionSign.cpp:
```diff
@@ -167,6 +167,22 @@ checkPayment(
+            if (tx_json[jss::DeliverMax] != tx_json[jss::Amount])
+                return RPC::make_error(
+                    rpcINVALID_PARAMS,
+                    "Cannot specify both 'Amount' and 'DeliverMax'");
```
Thanks, will change it.

Bronek commented on src/test/rpc/Transaction_test.cpp:
```diff
@@ -692,6 +693,63 @@ class Transaction_test : public beast::unit_test::suite
+            std::ostringstream ssExpected;
+            ssExpected << *memberIt;
+            BEAST_EXPECTS(
+                received == *memberIt,
```
testing equality of JSON fields listed in `expected` set, while allowing extra fields.

ckeshava commented on src/ripple/json/MultivarJson.h:
```diff
@@ -0,0 +1,109 @@
+struct MultivarJson
+{
+    std::array<Json::Value, Size> val;
+    constexpr static std::size_t size = Size;
```
If you like the suggestion, I've made the necessary changes at the tip [here](https://github.com/ckeshava/rippled/tree/bronek-sizeParam)

Bronek replied:
It's a good habit to expose key parameters with which a template class was instantiated. It's `constexpr static` , so it does not cost anything at all. I am going to stick to it - there's plenty of examples of this done e.g. in standard library

ckeshava commented on src/ripple/net/impl/InfoSub.cpp:
```diff
@@ -136,4 +136,17 @@ InfoSub::getRequest()
+unsigned int
+InfoSub::getApiVersion() const noexcept
+{
+    assert(apiVersion_ > 0);
```
Since `decltype(apiVersion_)` is an `unsigned int`, is this `assert` needed?

Bronek replied:
yes; 0 means that it was uninitialised.

arihantkothari reviewed (approved):
lgtm 🚀

ckeshava commented on src/ripple/app/misc/NetworkOPs.cpp:
```diff
@@ -599,10 +601,12 @@ class NetworkOPsImp final : public NetworkOPs
         bool validated,
-        std::shared_ptr<ReadView const> const& ledger);
+        std::shared_ptr<ReadView const> const& ledger,
+        std::optional<std::reference_wrapper<TxMeta const>> meta,
```
Since the `meta` parameter is only being read, its not being written into -- I attempted to remove the `std::reference_wrapper` [tip of this branch](https://github.com/ckeshava/rippled/tree/noRefWrap)
The unit tests are passing, but let me know the merits of using ref wrapper

Bronek replied:
I use `reference_wrapper` because it's an equivalent of passing by reference, i.e. saves you copying an object (`TxMeta` contains a `std::vector<STObject>` encapsulated inside `STArray` - this could be potentially heavy to copy). Technically `reference_wrapper` is a safe wrapper for a non-owning pointer.

This parameter is optional, and you _cannot_ put a reference type inside `std::optional`, so the most natural parameter type `std::optional<TxMeta const&>` won't work. Your choices are either:
- express in code what your intent is (which is what `std::optional<std::reference_wrapper<...>>` does), or
- pass a raw pointer (which can be null)

I find passing raw pointers an anti-pattern because they have potentially so many uses (many of them bad) that one needs to wonder "why here ?" when reading code. Which of course slows people down and may lead potential troubleshooters in the future on a wild goose chase.

ckeshava replied:
ok this helps. thanks for the explanation

Bronek commented on src/ripple/app/misc/NetworkOPs.cpp:
```diff
@@ -3743,21 +3794,24 @@ NetworkOPsImp::addAccountHistoryJob(SubAccountHistoryInfoWeak subInfo)
-                            jvTx[jss::account_history_tx_first] = true;
-                            send(jvTx, false);
+                            jvTx.set(jss::account_history_tx_first, true);
+                            sendMultiApiJson(jvTx, false);
```
We did not do logging before. Logging on networking operations is not a great idea, since it can potentially flood the logs e.g. on a persistent network failure. It's better to have metrics of this kind, but we did not have one here either. If we want it, it would belong to a different PR.

ximinez commented on src/ripple/json/MultivarJson.h:
```diff
@@ -48,11 +48,50 @@ struct MultivarJson
+
+NOTE:
+
+The more different JSON formats we support, the more CPU cycles we need to
+pre-build JSON for different API versions e.g. when publishing streams to
+`subscribe` clients. Hence it is desirable to keep MultiApiJson small and
+instead fully deprecate and remove support for old API versions. For example, if
+we removed support for API version 1 and added a different format for API
+version 3, the `apiVersionSelector` would change to
+`static_cast<std::size_t>(apiVersion > 2)`
```
No need to change code or comments, just something to think about.

Another possible solution is lazy loading, where we build a "base" version of the object, then provide a function to make the necessary per-apiVersion changes. When a new version is `select`ed, it'll copy and modify base object and store the result in the array.

Bronek replied:
Yeah I want to leave some guide there, so ppl do not have idea that we must always maintain some specific version for sake of this array.

ximinez commented on src/test/rpc/TransactionEntry_test.cpp:
```diff
@@ -164,13 +166,32 @@ class TransactionEntry_test : public beast::unit_test::suite
+                     memberIt++)
+                {
+                    auto const name = memberIt.memberName();
+                    BEAST_EXPECT(resIndex[jss::tx_json].isMember(name));
+                    auto const received = resIndex[jss::tx_json][name];
+                    std::ostringstream ssReceived;
+                    ssReceived << received;
+                    std::ostringstream ssExpected;
+                    ssExpected << *memberIt;
+                    BEAST_EXPECTS(
+                        received == *memberIt,
+                        txhash + " contains \n\"" + name + "\": "  //
+                            + ssReceived.str()                     //
+                            + "but expected "                      //
+                            + ssExpected.str());
```
I don't see this changed in the latest version.

Bronek replied:
Ouch, thanks for checking ! Done now.

Bronek commented on src/ripple/json/MultivarJson.h:
```diff
@@ -0,0 +1,109 @@
+    constexpr static std::size_t size = Size;
+
+    Json::Value const&
+    select(auto&& selector) const
```
For `select`: do not want to implicitly invoke any of the builtin conversion to `std::size_t` when the user lambda returns e.g. an `int`. IMO that's rather important because `std::size_t` is unsigned.
For `set`: do not want compilation error when Json element cannot be assigned from the parameter; would rather drop `set` from the overload set instead. Not that important, but it's SFINAE-friendly design

ckeshava replied:
I see, what about:
```
    Json::Value const&
    select(auto selector) const requires std::same_as(decltype(s), std::size_t) // no implicit conversion
    {
        return val[selector];
    }
    // would we still need a callable for selector?
```

Regarding your concern about the `set` method, all `Json::Value`'s should be assignable right? In what cases would we get a compilation error?

Bronek replied:
Re. design choice in `select` - passing a functor (or lambda) allows for improved code cohesion. I make use of it in `apiVersionSelector` - as you will notice, this makes coupling between index and api version more explicit and limited to one file. Well, in principle - I could make a better use of it.

Re. `set` , the error I meant was from assignment to `Json::Value` from some type which cannot be assigned to it, like for example demonstrated in unit tests `std::nullopt`

ckeshava replied:
Hmm alright, ok

Bronek commented on src/ripple/json/MultivarJson.h:
```diff
@@ -0,0 +1,109 @@
+    select(auto&& selector) const
+        requires std::same_as<std::size_t, decltype(selector())>
+    {
+        return val[selector()];
```
Good idea, will do that

ckeshava commented on src/test/rpc/Transaction_test.cpp:
```diff
@@ -692,6 +693,64 @@ class Transaction_test : public beast::unit_test::suite
+                std::ostringstream ssReceived;
+                ssReceived << received;
+                std::ostringstream ssExpected;
+                ssExpected << *memberIt;
```
You can do away with the `operator<<` with `Json::Value`'s `to_string` overload. It reduces the number of lines. If you like it, here's the [change](https://github.com/ckeshava/rippled/tree/json2Str)

Bronek replied:
Oh cool - did not know that!

Bronek commented on src/test/json/MultivarJson_test.cpp:
```diff
@@ -0,0 +1,278 @@
+                };
+            }(subject));
+            static_assert([](auto&& v) {
+                return requires
```
yeah learned that trick last year, esp. good for testing of interface (overloads etc.) in unit tests.

ximinez reviewed (changes requested):
Just a couple of minor things.

BTW, please stop force pushing. I'd rather have 20 tweak commits than have to figure out what changed since the last time I looked.

Bronek commented on src/ripple/app/misc/NetworkOPs.cpp:
```diff
@@ -3154,7 +3149,26 @@ NetworkOPsImp::transJson(
+    MultiApiJson multiObj({jvObj, jvObj});
+    static_assert(MultiApiJson::size == 2);
+    static_assert(RPC::apiMinimumSupportedVersion == 1);
+    for (unsigned apiVersion = 1, lastIndex = MultiApiJson::size;
+         apiVersion <= 2;
+         ++apiVersion)
```
You are right; also your comment prompted me to change the `static_assert` , so they are applied on a slightly higher abstraction level.

Bronek commented:
@ximinez thanks for very generous input - I addressed all your comments. I added one more test to 'MultivarJson_test.cpp` which required an extra line for levelization, but I think that should be OK.

arihantkothari commented on src/ripple/app/misc/NetworkOPs.cpp:
```diff
@@ -3136,7 +3149,31 @@ NetworkOPsImp::transJson(
+    static_assert(
+        apiVersionSelector(RPC::apiBetaVersion)() + 1  //
+        == MultiApiJson::size);
+    for (unsigned apiVersion = RPC::apiMinimumSupportedVersion,
```
For the intentions of looping over the apiVersions from `minimumSupported` to `beta`. I actually made a helper function long time back - may/ may not be useful :) 
https://github.com/XRPLF/rippled/pull/4611

ximinez replied:
That function is cool, but as written, it's only for tests.

arihantkothari replied:
oh shoot, my bad. I did not notice this was not for tests 😆

ximinez replied:
No worries

ximinez commented on Builds/levelization/results/ordering.txt:
```diff
@@ -135,6 +135,7 @@ test.csf > ripple.protocol
 test.csf > test.jtx
 test.json > ripple.beast
 test.json > ripple.json
+test.json > ripple.rpc
```
Adding lines to the ordering file is great, as long as it follows the existing rules (which this does). It would be a problem if the line was added to `loops.txt`, or if the dependency didn't "fit" the existing order.

intelliot commented:
Suggested commit message:
```
APIv2(DeliverMax): add alias for Amount in Payment transactions (#4733)

Using the "Amount" field in Payment transactions can cause incorrect
interpretation. There continue to be problems from the use of this
field. "Amount" is rarely the correct field to use; instead,
"delivered_amount" (or "DeliveredAmount") should be used.

Rename the "Amount" field to "DeliverMax", a less misleading name. With
api_version: 2, remove the "Amount" field from Payment transactions.

- Input: "DeliverMax" in `tx_json` is an alias for "Amount"
  - sign
  - submit (in sign-and-submit mode)
  - submit_multisigned
  - sign_for
- Output: Add "DeliverMax" where transactions are provided by the API
  - ledger
  - tx
  - tx_history
  - account_tx
  - transaction_entry
  - subscribe (transactions stream)
- Output: Remove "Amount" from API version 2

Fix #3484

Fix #3902
```

Bronek commented:
This is ready for merge

zgrguric commented:
Is "Amount" removed from output on Payment transactions?

ximinez commented:
> Is "Amount" removed from output on Payment transactions?

Only in API v2.
